pkg/trace/sampler/exception_sampler.go

```diff
@@ -0,0 +1,207 @@
+	defaultTTL = 2 * time.Minute
+	// priorityTTL allows to blacklist p1 spans that are sampled entirely, for this period.
+	priorityTTL = 10 * time.Minute
+	// ttlRenewalPeriod specifies the frequency at which we will upload cached entries.
+	ttlRenewalPeriod = 1 * time.Minute
+	// exceptionSamplerTPS traces per second allowed by the rate limiter.
+	exceptionSamplerTPS = 5
+	// exceptionSamplerBurst sizes the token store used by the rate limiter.
+	exceptionSamplerBurst = 50
+	exceptionKey          = "_dd.exception"
+)
+
+// ExceptionSampler samples traces that are not caught by the Priority sampler.
+// It ensures that we sample traces for each combination of
+// (env, service, name, resource, error type, http status) seen on a top level or measured span
+// for which we did not see any span with a priority > 0 (sampled by Priority).
+// The resulting sampled traces will likely be incomplete.
+type ExceptionSampler struct {
+	mu      sync.RWMutex
+	limiter *rate.Limiter
+	seen    map[Signature]*seenSpans
+
+	tickStats *time.Ticker
+	hits      int64
+	misses    int64
+	shrinks   int64
+}
+
+// NewExceptionSampler returns a NewExceptionSampler that ensures that we sample combinations
+// of env, service, name, resource, http-status, error type for each top level or measured spans
+func NewExceptionSampler() *ExceptionSampler {
+	e := &ExceptionSampler{
+		limiter:   rate.NewLimiter(exceptionSamplerTPS, exceptionSamplerBurst),
+		seen:      make(map[Signature]*seenSpans),
+		tickStats: time.NewTicker(10 * time.Second),
+	}
+	go func() {
+		for range e.tickStats.C {
+			e.report()
+		}
+	}()
+	return e
+}
+
+// Add samples a trace and returns true if trace was sampled (should be kept)
+func (e *ExceptionSampler) Add(env string, root *pb.Span, t pb.Trace) (sampled bool) {
+	return e.add(time.Now(), env, root, t)
+}
+
+func (e *ExceptionSampler) add(now time.Time, env string, root *pb.Span, t pb.Trace) (sampled bool) {
+	if priority, ok := GetSamplingPriority(root); priority > 0 && ok {
+		e.handlePriorityTrace(now, env, t)
+		return false
+	}
+	return e.handleTrace(now, env, t)
+}
+
+// Stop stops reporting stats
+func (e *ExceptionSampler) Stop() {
+	e.tickStats.Stop()
+}
+
+func (e *ExceptionSampler) handlePriorityTrace(now time.Time, env string, t pb.Trace) {
+	expire := now.Add(priorityTTL)
+	for _, s := range t {
+		if !traceutil.HasTopLevel(s) && !traceutil.IsMeasured(s) {
+			continue
+		}
+		e.addSpan(expire, env, s)
+	}
+}
+
+func (e *ExceptionSampler) handleTrace(now time.Time, env string, t pb.Trace) bool {
+	var sampled bool
+	expire := now.Add(defaultTTL)
+	for _, s := range t {
+		if !traceutil.HasTopLevel(s) && !traceutil.IsMeasured(s) {
+			continue
+		}
+		if !sampled {
+			sampled = e.sampleSpan(now, env, s)
+			continue
+		}
+		e.addSpan(expire, env, s)
+	}
+	return sampled
+}
+
+func (e *ExceptionSampler) addSpan(expire time.Time, env string, s *pb.Span) {
```

Comment:
Nit: some extra comments around all these internal methods (add / addSpan/ handle(Priority)Trace could help understand the logic.
At the end of the day, it's about keeping track of all past span signatures and keeping the trace only if it has been a long time we haven't seen it. Here I needed to read the whole file to get to understand it.